File: sentimental-credit/credit.py
import cs50
import re
import math

# Ask the user for their credit/debit card number
cardstr = cs50.get_string("Number: ")
cardint = int(cardstr)

# calculate checksum
sumOdd = 0
sumEven = 0
inProcessCardInt = cardint
for i in range(len(cardstr)):
    # if its in a odd position counting from the back
    if (i % 2 == 0):
        sumOdd += (inProcessCardInt % 10)
    # if its an even position counting from the back
    else:
        moduloByTwo = 2 * (inProcessCardInt % 10)
        sumEven += (moduloByTwo % 10)
        sumEven += (math.floor(moduloByTwo / 10) % 10)
    inProcessCardInt = math.floor(inProcessCardInt / 10)

checksum = sumOdd + sumEven

# if checksum does not end with 0 or is divisible by 10
if ((checksum % 10) != 0):
    # print INVALID
    print("INVALID")

# else check first or first two characters
else:
    # All American Express numbers start with 34 or 37; 15-digit numbers
    if (re.findall(r"\A34", cardstr) == ['34'] or re.findall(r"\A37", cardstr) == ['37']) and len(cardstr) == 15:
        # print AMEX
        print("AMEX")
    # most MasterCard numbers start with 51, 52, 53, 54, or 55, 16-digit numbers,
    elif (re.findall(r"\A[5][1-5]", cardstr) != []) and (len(cardstr) == 16):
        # print MASTERCARD
        print("MASTERCARD")
    # all Visa numbers start with 4. 13- and 16-digit numbers.
    elif re.findall(r"\A4", cardstr) == ['4'] and (len(cardstr) == 16 or len(cardstr) == 13):
        # print VISA
        print("VISA")
    # fail-safe in case its from another bank
    else:
        # print INVALID
        print("INVALID")

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // Ask the user for their credit/debit card number
    const long cardNumber = get_long("Number: ");

    // Declare and initialize variables for processing
    // Variable for checkSum
    long checkSumCardNumber = cardNumber;
    int sumEven = 0;
    int sumOdd = 0;
    // Variable to get the first two numbers
    long firstTwo = cardNumber;
    // Variable to count the amount of digits in the card number
    int digits = 0;

    // Checksum

    // Calculate the sum of even and odd positions in the card number
    // starting the count from the end
    for (int i = 0; checkSumCardNumber > 0; i++)
    {
        // obtain the last number using the modulo operation
        int modulo = checkSumCardNumber % 10;

        // divide by 10 so that the number "looses" its last number and its ready for next iteration
        checkSumCardNumber /= 10;

        // for odd positions in the card number:
        if ((i % 2) == 0)
        {
            sumOdd += modulo;
        }

        // for even positions in the card number:
        else
        {
            int moduloByTwo = modulo * 2;

            // Loop in case it has two figures
            while (moduloByTwo > 0)
            {
                sumEven += (moduloByTwo % 10);
                moduloByTwo /= 10;
            }
        }
        digits = i + 1;
    }

    // Add sumOdd and sumEven
    const int checkSum = sumOdd + sumEven;

    // Check if checkSum ends in 0 (is divisible by 10)
    if (checkSum % 10 != 0)
    {
        printf("INVALID\n");
    }
    else
    {
        // Find the first two digits
        while (firstTwo > 100)
        {
            firstTwo /= 10;
        }

        // All American Express numbers start with 34 or 37; 15-digit numbers
        if ((firstTwo == 34 || firstTwo == 37) && digits == 15)
        {
            printf("AMEX\n");
        }
        // most MasterCard numbers start with 51, 52, 53, 54, or 55, 16-digit numbers,
        else if ((firstTwo == 51 || firstTwo == 52 || firstTwo == 53 || firstTwo == 54 ||
                  firstTwo == 55) &&
                 digits == 16)
        {
            printf("MASTERCARD\n");
        }
        // all Visa numbers start with 4. 13- and 16-digit numbers.
        else if ((firstTwo / 10 == 4) && (digits == 16 || digits == 13))
        {
            printf("VISA\n");
        }
        // fail-safe in case its from another bank
        else
        {
            printf("INVALID\n");
        }
    }
}
